Add rating dialog & other fixes

ronits-wework opened the pull request:
Added an add rating dialog
Update restaurants and rating after updates to them
Fixed some styling

ronits-wework committed:
Added rating dialog

ronits-wework committed:
Fixed stars rating filter styling

ronits-wework committed:
Fetch restaurants again after adding a restaurant

ronits-wework committed:
Disabled submit button while adding review/restaurant
Fixed rating update bug
Styled restaurant card

GeReV commented on app/javascript/bundles/RestaurantsApp/components/AddRestaurant.jsx:
```diff
@@ -62,6 +65,7 @@ export default class AddRestaurantForm extends React.Component {
 
             const data = JSON.stringify(restaurant);
+            const onAdd = this.props.onAdd;
             const onClose = this.props.onClose;
```
Check this out: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment

TL;DR:
```js
const {
  onAdd,
  onClose,
} = this.props;
```

GeReV commented on app/javascript/bundles/RestaurantsApp/components/AddRestaurant.jsx:
```diff
@@ -62,6 +65,7 @@ export default class AddRestaurantForm extends React.Component {
             };
 
             const data = JSON.stringify(restaurant);
```
`fetch` might be able to handle non-stringified data (but I'm not certain).

ronits-wework replied:
Looks like it needs the stringify

GeReV commented on app/javascript/bundles/RestaurantsApp/components/AddRestaurant.jsx:
```diff
@@ -51,6 +53,7 @@ export default class AddRestaurantForm extends React.Component {
     submit(model) {
         this.setState({isFormSubmitted: true});
         if (this.state.isFormValid) {
+            this.setState({isAddingRestaurant: true});
```
For something like this I'd consider a more generic name, like `isSubmitting`.

It would make more sense if you ever want to extract your form handling code to a separate component.

GeReV commented on app/javascript/bundles/RestaurantsApp/components/AddReview.jsx:
```diff
@@ -0,0 +1,114 @@
+        return {
+            isFormValid: false
+        }
+    }
```
This method is probably redundant. You have something like this in the constructor.

GeReV commented on app/javascript/bundles/RestaurantsApp/components/AddReview.jsx:
```diff
@@ -0,0 +1,114 @@
+        this.setState({
+            isFormValid: false
+        });
+    }
```
This is totally valid, but consider merging them:
```js
setValid(valid = true) {
  this.setState({
    isFormValid: valid
  });
}

// Optional:
setInvalid() {
  this.setValid(false);
}
```

GeReV commented on app/javascript/bundles/RestaurantsApp/components/AddReview.jsx:
```diff
@@ -0,0 +1,114 @@
+
+    render() {
+        let formClass = this.state.isFormSubmitted ? "submitted" : "not-submitted";
+        formClass += " form";
```
There's a great library to help with class name handling, that's how Facebook originally did their classname handling (probably still do):
https://github.com/JedWatson/classnames

```js
import cx from 'classnames';

const className = cx('static-class', foo, /*...,*/ {
  bar: value, // Will render if value is truthy.
  // ...
});
```

GeReV commented on app/javascript/bundles/RestaurantsApp/components/AddReview.jsx:
```diff
@@ -0,0 +1,114 @@
+            <div className="add-review custom-form">
+                <h3 className="form-header">Rate {this.props.restaurant.name}</h3>
+                <Formsy.Form onSubmit={this.submit} onValid={this.setValid} onInvalid={this.setInvalid}
+                             className={formClass}>
```
Suggestion, format something like this like so:
```jsx
<Component
  attr={...}
  attr2={...}
  ...
>
```

GeReV commented on app/javascript/bundles/RestaurantsApp/components/RestaurantCard.jsx:
```diff
@@ -10,38 +25,93 @@ export default class RestaurantCard extends React.Component {
+                            </div>
+                            <div className="rest-sub-section">
+                                <div className="rest-sub-item">
+                                    {restaurant.kosher === true && (
```
Is it necessary to test that it's `true`? You could probably omit the `=== true` part.

GeReV commented on app/javascript/bundles/RestaurantsApp/components/RestaurantsApp.jsx:
```diff
@@ -106,6 +91,32 @@ export default class RestaurantsApp extends React.Component {
+            }).then((restaurants) => {
+            this.setState({
+                    restaurants
+                }, this.filterRestaurants
```
Nice use of the callback!

GeReV commented on app/javascript/bundles/RestaurantsApp/components/AddReview.jsx:
```diff
@@ -0,0 +1,114 @@
+                    method: "POST",
+                    body: data,
+                    headers: {
+                        'Accept': 'application/json, text/plain, */*',
```
Could probably just omit this bit.

GeReV commented on app/javascript/bundles/RestaurantsApp/components/AddReview.jsx:
```diff
@@ -0,0 +1,114 @@
+                    onAdd();
+                })
+                .catch(function (error) {
+                    console.log(error);
```
You can also use `console.error` and `console.warn`.

GeReV commented on app/javascript/bundles/RestaurantsApp/components/RestaurantsApp.jsx:
```diff
@@ -106,6 +91,32 @@ export default class RestaurantsApp extends React.Component {
+        fetch("/restaurants.json")
+            .then(res => {
+                return res.json();
+            }).then((restaurants) => {
```
You can simplify this to `then(res => res.json())` for brevity.

GeReV reviewed (approved):
Some suggestions, but looks good.

ronits-wework committed:
Changed code review suggestions

ronits-wework committed:
Refactor StarsRatingFilter

ronits-wework committed:
Began google maps integration

ronits-wework committed:
Add restaurant card to marker hover

ronits-wework committed:
Set map center according to user location

ronits-wework committed:
Separate page header

ronits-wework committed:
Added a clear all filters button
Started fixing filters to clear properly

ronits-wework committed:
Finish fixing filters to clear properly

ronits-wework committed:
Add the option for multiple cuisine types

ronits-wework committed:
Added AddressInput to AddRestaurantForm
Loaded google api places for PlacesAutocomplete during loading of RestaurantsMap, to avoid a double load

ronits-wework committed:
Moved google api loading to main app, to be able to support multiple map components

ronits-wework committed:
Added animation to map marker hover

ronits-wework committed:
Zoom into restaurant when clicked, display as hover when card from list is hovered
Add scrolling to restaurants

ronits-wework committed:
Fixed rating bug, added some styles

ronits-wework committed:
Fixed some styles

ronits-wework committed:
Added delivery times and reviews to initial data

ronits-wework committed:
Added a wait cursor when submitting a form

ronits-wework committed:
Restyle header

ronits-wework committed:
Restyle header

ronits-wework committed:
Added a disabled state to clear filters button

ronits-wework committed:
Restyle header

ronits-wework committed:
Restyle header

GeReV commented on app/assets/stylesheets/add_restaurant.scss:
```diff
@@ -5,4 +5,8 @@
+
+  .wait {
+    cursor: wait;
+  }
```
This can be a global class :P

GeReV commented on app/assets/stylesheets/map_marker.scss:
```diff
@@ -0,0 +1,98 @@
+@mixin hover-animation() {
+  -webkit-transition: $hover-animation-duration; /* Safari */
+  transition: $hover-animation-duration;
+  transition-timing-function: ease-in;
```
I think you're missing the properties to transition.
See: https://developer.mozilla.org/en-US/docs/Web/CSS/transition

GeReV commented on app/javascript/bundles/Forms/components/AddressInput.jsx:
```diff
@@ -0,0 +1,50 @@
+
+    constructor(props) {
+        super(props);
+    }
```
I believe that's redundant, if there's no additional code here.

GeReV commented on app/javascript/bundles/Forms/components/AddressInput.jsx:
```diff
@@ -0,0 +1,50 @@
+    };
+
+    render() {
+        const className = "";
```
Seems unnecessary.

GeReV commented on app/javascript/bundles/Forms/components/AddressInput.jsx:
```diff
@@ -0,0 +1,50 @@
+        .then(results => getLatLng(results[0]))
+        .then(latLng => { return {longitude: latLng.lng, latitude: latLng.lat}})
+        .catch(error => null);
+}
```
+1

GeReV commented on app/javascript/bundles/Map/components/RestaurantMapMarker.jsx:
```diff
@@ -0,0 +1,72 @@
+    render() {
+        const markerClass = cx('map-marker', {
+            'selected': this.props.isSelected,
+        });
```
😄

GeReV replied:
No need for quotes around `selected`, since it's a valid key name.
You would need them if it contained non-word characters, say, `foo-selected`.

GeReV commented on app/javascript/bundles/Map/components/RestaurantsMap.jsx:
```diff
@@ -0,0 +1,75 @@
+
+        this.state = {
+            center: [32.078668, 34.781235],
+            zoom: 14,
```
Looks like good candidates to extract to consts.

GeReV commented on app/javascript/bundles/Map/components/RestaurantsMap.jsx:
```diff
@@ -0,0 +1,75 @@
+                        lng={restaurant.longitude}
+                        text={restaurant.name}
+                        restaurant={restaurant}
+                        isSelected={(this.props.selectedRestaurant ? (this.props.selectedRestaurant.id === restaurant.id) : false)}
```
I think you can just put the first half.
If there's no `selectedRestaurant` set, no ID would be equal to it ever.

GeReV commented on app/javascript/bundles/Map/components/RestaurantsMap.jsx:
```diff
@@ -0,0 +1,75 @@
+                        isSelected={(this.props.selectedRestaurant ? (this.props.selectedRestaurant.id === restaurant.id) : false)}
+                        onRestaurantsChange={this.props.onRestaurantsChange}
+                    />)
+                })}
```
I think a better formatting would be:
```js
{this.props.restaurants.map((restaurant) => (
  <RestaurantMapMarker
    key={restaurant.id}
    lat={restaurant.latitude}
    lng={restaurant.longitude}
    text={restaurant.name}
    restaurant={restaurant}
    isSelected={(this.props.selectedRestaurant ? (this.props.selectedRestaurant.id === restaurant.id) : false)}
    onRestaurantsChange={this.props.onRestaurantsChange}
  />
))}
```

GeReV commented on app/javascript/bundles/RestaurantsApp/components/AddRestaurant.jsx:
```diff
@@ -53,49 +50,65 @@ export default class AddRestaurantForm extends React.Component {
+            const {
+                onAdd,
+                onClose
+            } = this.props;
```
I usually put this at the top of the class, just because it's easier to find when skimming through the code.

GeReV commented on app/javascript/bundles/RestaurantsApp/components/Restaurants.jsx:
```diff
@@ -15,13 +15,22 @@ export default class Restaurants extends React.Component {
+                    className="restaurant-card-wrapper"
+                    onClick={() => this.props.onRestaurantClick(restaurant)}
+                    onMouseEnter={() => this.props.onRestaurantEnter(restaurant)}
+                    onMouseLeave={() => this.props.onRestaurantLeave(restaurant)}
```
I'm guessing this is used for showing/hiding some elements?
This is something I would try to do with only CSS as much as possible, but it's not imperative.

GeReV commented on app/javascript/bundles/RestaurantsApp/components/RestaurantsApp.jsx:
```diff
@@ -60,6 +67,11 @@ export default class RestaurantsApp extends React.Component {
+        this.loadGoogleMap = this.loadGoogleMap.bind(this);
+        this.onRestaurantClick = this.onRestaurantClick.bind(this);
+        this.onRestaurantEnter = this.onRestaurantEnter.bind(this);
+        this.onRestaurantLeave = this.onRestaurantLeave.bind(this);
```
Seems to me like a good indicator there's some refactoring in order. :)

GeReV replied:
For example, there's no apparent (that I can see) need for `loadGoogleMap` and `loadJS` to be in the component.
You can simply extract them out of the component into their own functions, or even their own file.

ronits-wework replied:
I refactored according to this article: https://medium.freecodecamp.org/react-binding-patterns-5-approaches-for-handling-this-92c651b5af56
Using the 5th method, defining the function as an arrow function

GeReV commented on app/javascript/bundles/RestaurantsApp/components/RestaurantsApp.jsx:
```diff
@@ -118,6 +181,33 @@ export default class RestaurantsApp extends React.Component {
+        );
+
+        return !noFilters;
+    }
```
I don't know what would be the best way to implement this, but looks to me like this could be written in a more scalable manner (think of, for example, a map of filter name to function, for example), and an `activeFilters` object.
See @yonidor's comment here: https://github.com/NoyaArie/AppWeEat/pull/5#discussion_r143509363

We can talk about possible implementations if you'd like, but you can consider this out of scope.

GeReV commented on app/javascript/bundles/RestaurantsApp/components/RestaurantsApp.jsx:
```diff
@@ -266,14 +367,35 @@ export default class RestaurantsApp extends React.Component {
-                        onRestaurantsChange={this.onRestaurantsChange}
+                    <img
+                        className={clearBtnClass}
+                        onClick={this.clearFilters}
```
I, personally, don't like `onClick`s on non-button or link elements...
But that's just me.

GeReV commented on app/javascript/bundles/RestaurantsApp/components/RestaurantsApp.jsx:
```diff
@@ -266,14 +367,35 @@ export default class RestaurantsApp extends React.Component {
+                                if (restaurant.latitude === null || restaurant.longitude === null) {
+                                    return false;
+                                }
+                                return true;
```
This can be written in a much simpler manner. Just saying.

GeReV commented on app/javascript/bundles/StarsRating/components/StarsRating.jsx:
```diff
@@ -77,8 +77,8 @@ export default class StarsRating extends React.Component {
-            this.props.ratingCallback(newRating);
+        if (this.props.onRating) {
+            this.props.onRating(newRating);
         }
```
There's a nice way of avoiding a check like this; you can make `onRating` be a "no-op" function by default.
For example:
```js
static defaultProps = {
  onRating: function() {},
}
```

That way you don't have to check if it exists and just call it.

You can also define it in some `noop.js` file and export it, if you use it in more than one place.

There are a bunch of libraries, like Lodash, which provide a No-Op function that you can import, if you happen to already use one.
(Not that there's any real need to _import_ an empty function...)

But often that goes over peoples' heads and they need it to be explained to them...
Alternatively, you can define an `onRating` method on the component that will perform this check:
```js
onRating(newRating) {
  if (this.props.onRating) {
    this.props.onRating(newRating);
  }
}

// And just call:
this.onRating(newRating);
```

GeReV commented on lib/tasks/import_restaurants.rake:
```diff
@@ -4,18 +4,27 @@ namespace :import_restaurants_csv do
         cuisine = CuisineType.find_by_cuisine(row[:cuisine])
         RestaurantCuisine.create!(cuisine_type_id: cuisine.id, restaurant_id: restaurant.id)
+
+        @reviews_num = Faker::Number.between(0, 5)
```
I think a simple random would do...
```ruby
Random.rand(0..5)
```

GeReV commented on lib/tasks/import_restaurants.rake:
```diff
@@ -4,18 +4,27 @@ namespace :import_restaurants_csv do
+          Review.create!(name: Faker::Name.name, rating: Faker::Number.between(0, 3),
+                         comment: Faker::Lorem.sentence, restaurant_id: restaurant.id);
+          @i += 1
+        end
```
Suggestion:
```ruby
@reviews_num.times do
  Review.create!(...)
end
```

ronits-wework committed:
Implement some changes from code review

ronits-wework committed:
Code review change - separate google map resource loading to a separate file

ronits-wework committed:
Code review change - refactor function bindings in constructor, use arrow functions instead

ronits-wework committed:
Code review change - begin refactoring filters to be more generic

ronits-wework committed:
Code review change - continue refactoring filters, separate classes to files

ronits-wework committed:
Fixed bug in map - after navigating through the map, centering on a restaurant stopped working
